src/net/neighbors/walk.rs

```diff
@@ -0,0 +1,2342 @@
+        }
+        ret
+    }
+
+    /// Given a neighbor we tried to insert into the peer database, find one of the existing
+    /// neighbors it collided with.  Return its slot in the peer db.
+    fn find_replaced_neighbor_slot(
+        conn: &DBConn,
+        nk: &NeighborKey,
+    ) -> Result<Option<u32>, net_error> {
+        let mut slots = PeerDB::peer_slots(conn, nk.network_id, &nk.addrbytes, nk.port)
+            .map_err(net_error::DBError)?;
+
+        if slots.len() == 0 {
+            // not present
+            return Ok(None);
+        }
+
+        let mut rng = thread_rng();
+        slots.shuffle(&mut rng);
+
+        for slot in slots {
+            let peer_opt =
+                PeerDB::get_peer_at(conn, nk.network_id, slot).map_err(net_error::DBError)?;
+
+            match peer_opt {
+                None => {
+                    continue;
+                }
+                Some(_) => {
+                    return Ok(Some(slot));
+                }
+            }
+        }
+
+        Ok(None)
+    }
+
+    /// Add a neighbor or schedule it to be pinged since it's up for replacement.
+    /// Returns (was-new?, neighbor)
+    fn add_or_schedule_replace_neighbor<'a>(
+        &mut self,
+        tx: &mut DBTx<'a>,
+        block_height: u64,
+        naddr: &NeighborAddress,
+        peer_version: u32,
+        network_id: u32,
+        handshake: &HandshakeData,
+        db_data: Option<&StackerDBHandshakeData>,
+    ) -> Result<(bool, Neighbor), net_error> {
+        let mut neighbor_from_handshake =
+            Neighbor::from_handshake(tx, peer_version, network_id, handshake)?;
+
+        if Neighbor::from_neighbor_address(tx, network_id, block_height, naddr)?.is_some() {
+            test_debug!("already know about {:?}", naddr);
+            neighbor_from_handshake
+                .save_update(tx, db_data.map(|x| x.smart_contracts.as_slice()))?;
+
+            // seen this neighbor before
+            return Ok((false, neighbor_from_handshake));
+        }
+
+        debug!("new neighbor {:?}", &neighbor_from_handshake.addr);
+
+        // didn't know about this neighbor yet. Try to add it.
+        let added =
+            neighbor_from_handshake.save(tx, db_data.map(|x| x.smart_contracts.as_slice()))?;
+
+        if added {
+            // neighbor was new, and we had space to add it.
+            return Ok((true, neighbor_from_handshake));
+        }
+
+        // neighbor was new, but we don't have space to insert it.
+        // find and record a neighbor it would replace.
+        let replaced_neighbor_slot_opt =
+            NeighborWalk::find_replaced_neighbor_slot(tx, &neighbor_from_handshake.addr)?;
+        if let Some(slot) = replaced_neighbor_slot_opt {
+            // if this peer isn't allowed or denied, then consider
+            // replacing.  Otherwise, keep the local configuration's preference.
+            if !neighbor_from_handshake.is_denied() && !neighbor_from_handshake.is_allowed() {
```

Comment:
I fixed `is_allowed()` and `is_denied()` so that they're now mutually exclusive.  Now, this check only uses `.is_denied()`.

Reply:
I think this check just moved to `neighbors/db.rs:401`.

My confusion is why a new neighbor that is "allowed" should be treated the same as a new neighbor that is "denied". The if statement here makes it so that a replacement is only scheduled if the new neighbor is denied _or_ the new neighbor is allowed.

Reply:
Notions of "allowed" and "denied" in this context refer to local policy.  A node is "allowed" if the operator explicitly marked it as such in the PeerDB (e.g. through a config option), meaning that it should never be banned or blocked by the network.  Similarly, a node is "denied" if the operator explicitly marked it as such, or if it got banned and was not "allowed."

Under these terms, a node can be both "allowed" and "denied" through a bad local configuration, which really ought to never happen.  The point of this code was to avoid overriding the node operator's preferences -- if the node operator explicitly allowed or denied a node, we don't want to replace its record in the PeerDB (because we want those policies to stay that way).

Reply:
But isn't a neighbor only in this code path if it isn't already stored in the peer db? So is this a dead check (i.e., there's no way for a neighbor to return true for "allowed" and the code to get to this line, because it would have returned somewhere around the check `Neighbor::from_neighbor_address(&mut tx, preamble.network_id, block_height, naddr)?.is_some()`?

Reply:
No, because `is_allowed()` is a function of the `allowed` value in `Neighbor`, which is a UNIX timestamp.  A neighbor can be marked as allowed or denied until a certain time (in the latter case, we use the timestamp to permit bans to expire).  A neighbor can be loaded via `Neighbor::from_neighbor_address()` if its `allowed` value is in the past (meaning, `is_allowed()` is false).

Reply:
I'm sorry to keep circling around on this point, but I think I'm still pretty confused here.

Here is the start of the method we're talking about:

```rust
        let mut neighbor_from_handshake = Neighbor::from_handshake(
            &mut tx,
            preamble.peer_version,
            preamble.network_id,
            handshake,
        )?;
```

So at the start of the method, `neighbor_from_handshake` is initialized either (1) entirely from the handshake data or (2) from the db and then updated with the handshake data. My understanding, based on what you're saying, is that the _only_ way `is_allowed()` could be true is if the case was (2).

Immediately after that, the method then does this check:

```rust
        if Neighbor::from_neighbor_address(&mut tx, preamble.network_id, block_height, naddr)?
            .is_some()
        {
            test_debug!("already know about {:?}", naddr);
            neighbor_from_handshake
                .save_update(&mut tx, db_data.map(|x| x.smart_contracts.as_slice()))?;
            tx.commit()?;

            // seen this neighbor before
            return Ok((false, neighbor_from_handshake));
        }
```

`Neighbor::from_neighbor_address` returns `Some` iff the neighbor identified by (`preamble.network_id`, `naddr` ) is in the db. I believe that `naddr` and the (`addrbytes`, `port`) in `handshake` are the same (i.e., `naddr` is a reference to the neighbor that returned the `handshake` data). In which case, `Neighbor::from_neighbor_address` returns `Some` if `neighbor_from_handshake` was loaded from the db (i.e., case (2) above). If it returns `Some`, then the method _will return_ at this point.

So, if `neighbor_from_handshake.is_allowed()` is true, then `neighbor_from_handshake` was loaded from the db. If it was loaded from the db, then this if statement executes. If this if statement executes, the method returns before ever getting to the save or replace logic.

Reply:
> (2) from the db and then updated with the handshake data. My understanding, based on what you're saying, is that the only way is_allowed() could be true is if the case was (2).

This is not true.  `is_allowed()` can return false but the neighbor will be loaded.  This is because `neighbor_from_handshake` uses `PeerDB::get_peer()`, which doesn't look at allowed or denied policy.  `PeerDB::get_peer()` (and thus this method) will load a neighbor that is banned.

The `PeerNetwork` will not connect to a denied peer, however.

Reply:
But it doesn't matter if `is_allowed()` _could_ return false, the check in question is guarding against `is_allowed` being true:

```
            if !neighbor_from_handshake.is_denied() && !neighbor_from_handshake.is_allowed() {
                replacements.add_neighbor(
                    NeighborAddress::from_neighbor(&neighbor_from_handshake),
                    neighbor_from_handshake.clone(),
                    slot,
                );
            } else {
                unreachable!("My point is that this hypothetical case being guarded against seems unreachable.");
            }
```

Reply:
If either `is_denied()` or `is_allowed()` is true, then `neighbor_from_handshake` was loaded from the database (I'm not saying that the inverse is true: I understand that these methods can return false for data loaded from the database).

If `neighbor_from_handshake` was loaded from the database, then the method returns before ever performing that `!is_denied() && !is_allowed()` check, because it returns in the branch corresponding to the `test_debug!("already know about {:?}", naddr);`

Reply:
Oh, I see now.  Yes, it appears that this is a dead check, because as you say, if either `is_denied()` or `is_allowed()` is `true`, then the `allowed` and/or `denied` values must have been loaded from the DB since they'd be non-zero.  I'll remove the check.